### PR TITLE
Make it possible to notify the TrustManager of resumed sessions

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
Some trust manager implementations produce custom user principals and store them in the SSLSession. Resumed TLS sessions, however, don't always recover the stored contents. For instance, with TLSv1.3 stateless session resumption, our TLS implementations only store the peer certificate chain (or even just the leaf cert) in the session ticket. In such a case, the trust manager would like to be notified of the resumption, so that the peer principal can be recreated and stored in the session once again.

Modification:
Add a `ResumableX509ExtendedTrustManager` interface, with callbacks for resumed client and server sessions. Add infrastructure (the `ResumptionController`) to detect session resumption in an SSLEngine implementation agnostic way; if a TLS handshake completed without any calls to the TrustManager, then we've made a "stateless resumption" (hand-wave details) and should call the appropriate resume method. The JDK APIs don't provide a method to reliably discern if a session has been resumed, so we instead detect if the trust manager was called during the handshake. We do this by wrapping user-supplied trust managers that implement the `ResumableX509ExtendedTrustManager` interface. The wrapper will monitor calls to the trust manager and register the relevant `SSLEngine` in the controller — the engine is the object that the trust manager and the `SslHandler` both have access to, which reliably identifies the specific TLS session. Before finished the handshake, the `SslHandler` queries the `ResumptionController` to see if the trust manager has already been called, or if it needs to be notified of resumption. Since it's possible that the peer certificates have expired since the initial session, the trust manager gets another change to throw CertificateException.

Result:
Handlers that expect to find user principals in the authenticated SSLSessions no longer find empty sessions, provided they use a trust manager that correctly implements ResumableX509ExtendedTrustManager.

This is a draft to gage if the general approach is sane. There's still a few things that need to be done:

- [x] Javadocs.
- [x] Tests where the `resume*` methods throw `CertificateException`.
- [x] Tests for different transports.
- [x] Tests that use TCP FastOpen.
- [x] Tests that use startTls.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 1]
```diff
@@ -1943,14 +1955,36 @@ public void run() {
      * @return {@code true} if {@link #handshakePromise} was set successfully and a {@link SslHandshakeCompletionEvent}
      * was fired. {@code false} otherwise.
      */
-    private boolean setHandshakeSuccess() {
+    private boolean setHandshakeSuccess() throws SSLException {
         // Our control flow may invoke this method multiple times for a single FINISHED event. For example
         // wrapNonAppData may drain pendingUnencryptedWrites in wrap which transitions to handshake from FINISHED to
         // NOT_HANDSHAKING which invokes setHandshakeSuccess, and then wrapNonAppData also directly invokes this method.
+        SSLSession session = engine.getSession();
+        if (resumptionController != null && session.getPeerPrincipal() instanceof X500Principal) {
```
Comment (chrisvest) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1963]: Looks like we can't make this instanceof check, but have to accept that "peer not verified" may be thrown

**Comment by chrisvest** [comment]
Looks like Conscrypt on Windows is broken somehow.

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Great idea!

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 2]
```diff
@@ -1943,14 +1955,24 @@ public void run() {
...
+                if (resumptionController.validateResumeIfNeeded(engine) && logger.isDebugEnabled()) {
+                    logger.debug("{} Resumed and reauthenticated session (handshake_done={})", ctx.channel(),
+                            handshakePromise.isDone());
+                }
+            } catch (CertificateException e) {
+                throw new SSLException(e);
+            }
+        }
         final boolean notified;
         if (notified = !handshakePromise.isDone() && handshakePromise.trySuccess(ctx.channel())) {
```
Comment (chrisvest) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1974]: @normanmaurer One thing I'm wondering if whether the resumption block should be behind this handshake-is-done gate. The code implies that the handshake promise could already be completed at this point, but I can't think of a case where that would be true, or what it would mean.

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+final class ResumptionController {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(ResumptionController.class);
+    private static final Object ENTRY = new Object();
+    private final Map<SSLEngine, Object> confirmedValidations;
+    private final AtomicReference<ResumableX509ExtendedTrustManager> resumableTm;
+    private final boolean isClient;
+    private final boolean logResumptionFailure;
+
+    ResumptionController(boolean isClient, boolean logResumptionFailure) {
+        this.isClient = isClient;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:47]: actually we don't need `isClient` as we can use `engine.getUseClientMode()`

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        if (resumableTm.get() != null) {
+            confirmedValidations.remove(unwrapEngine(engine));
+        }
+    }
+
+    public boolean validateResumeIfNeeded(SSLEngine engine)
+            throws CertificateException, SSLPeerUnverifiedException {
+        ResumableX509ExtendedTrustManager tm;
+        boolean valid = engine.getSession().isValid();
+        if ((tm = resumableTm.get()) != null && valid) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:80]: nit: we could check `valid` first if we don't log as it is cheaper.

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    private static X509Certificate[] chainOf(Certificate[] peerCertificates) {
+        if (peerCertificates instanceof X509Certificate[]) {
+            //noinspection SuspiciousArrayCast
+            return (X509Certificate[]) peerCertificates;
+        }
+        X509Certificate[] chain = new X509Certificate[peerCertificates.length];
+        for (int i = 0; i < peerCertificates.length; i++) {
+            chain[i] = (X509Certificate) peerCertificates[i];
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:112]: Should we do an instanceof check here first and throw an `IlegalArgumentException` if the type not match ?

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        }
+        return chain;
+    }
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    private class X509ExtendedWrapTrustManager extends X509ExtendedTrustManager {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:118]:
```suggestion
    private static final class X509ExtendedWrapTrustManager extends X509ExtendedTrustManager {
```

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        private final X509ExtendedTrustManager trustManager;
+
+        X509ExtendedWrapTrustManager(X509ExtendedTrustManager trustManager) {
+            this.trustManager = ObjectUtil.checkNotNull(trustManager, "trustManager");
+        }
+
+        private void unsupported() throws CertificateException {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:125]: this could be static
```suggestion
        private static void unsupported() throws CertificateException {
```

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 2 above)
Comment (normanmaurer) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1974]: yeah sounds more robust to me... that said this would also mean that the promise would be notified before we validate which is a bit "odd" if people expect that all the data etc is restored.

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 8]
```diff
@@ -2027,6 +2049,9 @@ private void setHandshakeFailureTransportFailure(ChannelHandlerContext ctx, Thro
     }
 
     private void releaseAndFailAll(ChannelHandlerContext ctx, Throwable cause) {
+        if (resumptionController != null) {
+            resumptionController.remove(engine());
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:2053]: hmm... is this really what we want ? I think we only want to remove it if `isValid()` is false or if the exception is a `SslHandshakeException`

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport/src/test/java/io/netty/bootstrap/BootstrapTest.java [hunk 9]
```diff
@@ -362,6 +362,7 @@ public void testAsyncResolutionFailure() throws Exception {
         // Should fail with the UnknownHostException.
         assertThat(connectFuture.await(10000), is(true));
         assertThat(connectFuture.cause(), is(instanceOf(UnknownHostException.class)));
+        connectFuture.channel().closeFuture().await(10000);
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/bootstrap/BootstrapTest.java:365]: Seems like this is unrelated ? Should this be an extra PR ?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 2 above)
Comment (chrisvest) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1974]: I definitely want the resumption call to happen before the promise listeners are notified. Some frameworks look for the pipeline event, while others listen on the handshake promise. They need to behave the same in this regard...

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 8 above)
Comment (chrisvest) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java:2053]: Calls to this method are always associated with failing the handshake promise. The promise can be reinitialized by a renegotiation, though I'm not sure if that's a possibility after we get here.
If we get here from `setHandshakeFailureTransportFailure` then the exception won't be an `SslHandshakeException` for instance.

If we filter out too many remove calls, then the failure mode is that an engine will remain in the `confirmedValidations` and can prevent a `resume` call from happening on session resumption. If we filter out too few remove calls, then if somehow a `remove` call happened before a `validateResumeIfNeeded` call, then we'd call both a `check*` method and a `resume*` method on the trust manager.

The latter seems to be a possibility e.g. in this stretch of code https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L973-L997 because the remove is not guarded by changing the handshake promise state. Perhaps that's the real fix? That the `remove` call should follow `handshakePromise.tryFailure`?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 2 above)
Comment (chrisvest) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1974]: I've looked at two approaches here:

One where we add a `trySuccess` method that also takes a `Runnable` callback which will be run before the listeners, if we successfully change the state of the promise. To make this atomic, the runnable callback will have to be executed while holding the lock on the promise, so no listeners can be added or notified.

Another where we add a `trySuccess` method which returns a "lease" object iff we successfully set the result of the promise to this lease object. When a promise has a lease result, is not "done" and cannot be completed by anyone else. The lease object is runnable, and will set the result of the promise to the real result and notify the listeners, when run. This is more complicated, but we won't need to hold a lock on the promise for an extended period.

Comment (chrisvest) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: Oh, I realized that we _have_ to try calling the `resume` methods before completing the handshake promise, as otherwise we can't fail it if the trust manager throws an exception.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 10]
```diff
@@ -16,6 +16,7 @@
 package io.netty.util.concurrent;
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:16]: Can you tell me a bit more about these changes ?

Comment (chrisvest) [reply common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:16]: Oh, these are the remains of the changes discussed in https://github.com/netty/netty/pull/14358#discussion_r1774190996
They should be reverted.

**Review by normanmaurer — APPROVED** [review]

**Comment by chrisvest** [comment]
I think it's fine to not specifically test with startTls because https://github.com/netty/netty/issues/14373

**Comment by chrisvest** [comment]
Subclasses of `SocketSslSessionReuseTest` cover the TCP FastOpen case as well.

**Comment by chrisvest** [comment]
@normanmaurer I think this is good to go, assuming the tests are no longer flaky.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 11]
```diff
@@ -667,7 +667,7 @@ private void rethrowIfFailed() {
         if (cause == null) {
             return;
         }
-
+        cause.addSuppressed(new RuntimeException("tracer"));
```
Comment (chrisvest) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:670]: Temporary for debugging the flaky tests

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 12]
```diff
@@ -667,7 +667,7 @@ private void rethrowIfFailed() {
         if (cause == null) {
             return;
         }
-
+        ThrowableUtil.addSuppressed(cause, new RuntimeException("tracer"));
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:670]: TODO: Remove

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+import java.util.WeakHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509ExtendedTrustManager;
+
+final class ResumptionController {
+    private static final Object ENTRY = new Object();
+    private final Map<SSLEngine, Object> confirmedValidations;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:37]: Nit: we could also make this `Set<SSLEngine>` and just use `Collections.newSetFromMap(....)`

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    private final AtomicReference<ResumableX509ExtendedTrustManager> resumableTm;
+
+    ResumptionController() {
+        confirmedValidations = Collections.synchronizedMap(new WeakHashMap<SSLEngine, Object>());
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:41]: See above

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+        resumableTm = new AtomicReference<ResumableX509ExtendedTrustManager>();
+    }
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    public TrustManager wrapIfNeeded(TrustManager tm) {
+        if (tm instanceof ResumableX509ExtendedTrustManager) {
+            if (PlatformDependent.javaVersion() < 7 || !(tm instanceof X509ExtendedTrustManager)) {
+                throw new IllegalStateException(
+                        "Resumable trust manager must be a subclass of X509ExtendedTrustManager");
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:50]:
```suggestion
                        "ResumableX509ExtendedTrustManager implementation must be a subclass of X509ExtendedTrustManager, found: " +
                        (tm == null ? null : tm.getClass())););
```

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+        }
+    }
+
+    public boolean validateResumeIfNeeded(SSLEngine engine)
+            throws CertificateException, SSLPeerUnverifiedException {
+        ResumableX509ExtendedTrustManager tm;
+        boolean valid = engine.getSession().isValid();
+        if (valid && (tm = resumableTm.get()) != null) {
+            Certificate[] peerCertificates = engine.getSession().getPeerCertificates();
+            engine = unwrapEngine(engine);
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:72]: Can we add some comment to why we do the unwrapping ? I think it is because the "TrustManager" also only sees the unwrapped engine but it would be good to clarify

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    public TrustManager wrapIfNeeded(TrustManager tm) {
+        if (tm instanceof ResumableX509ExtendedTrustManager) {
+            if (PlatformDependent.javaVersion() < 7 || !(tm instanceof X509ExtendedTrustManager)) {
+                throw new IllegalStateException(
+                        "Resumable trust manager must be a subclass of X509ExtendedTrustManager");
+            }
+            if (!resumableTm.compareAndSet(null, (ResumableX509ExtendedTrustManager) tm)) {
+                throw new IllegalStateException("Only one TrustManager can be configured for resumed sessions");
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:53]:
```suggestion
                throw new IllegalStateException("Only one ResumableX509ExtendedTrustManager can be configured for resumed sessions");
```

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+        private final Map<SSLEngine, Object> confirmedValidations;
+
+        X509ExtendedWrapTrustManager(X509ExtendedTrustManager trustManager,
+                                     Map<SSLEngine, Object> confirmedValidations) {
+            this.trustManager = ObjectUtil.checkNotNull(trustManager, "trustManager");
+            this.confirmedValidations = confirmedValidations;
+        }
+
+        private static void unsupported() throws CertificateException {
+            throw new CertificateException("Resumable trust managers require the SSLEngine parameter");
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:125]: Why not throw `UnsupportedOperationException` ?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 19]
```diff
@@ -1943,14 +1951,26 @@ public void run() {
...
+    private boolean setHandshakeSuccess() throws SSLException {
         // Our control flow may invoke this method multiple times for a single FINISHED event. For example
         // wrapNonAppData may drain pendingUnencryptedWrites in wrap which transitions to handshake from FINISHED to
         // NOT_HANDSHAKING which invokes setHandshakeSuccess, and then wrapNonAppData also directly invokes this method.
+        final SSLSession session = engine.getSession();
+        if (resumptionController != null && !handshakePromise.isDone()) {
+            try {
+                if (resumptionController.validateResumeIfNeeded(engine) && logger.isDebugEnabled()) {
+                    logger.debug("{} Resumed and reauthenticated session (handshake_done={})", ctx.channel(),
+                            handshakePromise.isDone());
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1963]: wouldn't this be always `false` ?

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 20]
```diff
@@ -3808,6 +3845,257 @@ public void checkServerTrusted(java.security.cert.X509Certificate[] chain,
...
+            public void messageReceived(ChannelHandlerContext ctx, ByteBuf msg) throws Exception {
+                SslHandler sslHandler = ctx.pipeline().get(SslHandler.class);
+                SSLEngine engine = sslHandler.engine();
+                logger.debug("Client message received: {} ({}) {} {}",
+                        engine.getSession(), engine.getHandshakeStatus(), isSessionReused(engine),
+                        Arrays.toString(engine.getSession().getValueNames()));
+                Object value = engine.getSession().getValue("key");
+                clientSessionValues.put((String) value);
+                msg.release();
+            }
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java:3870]:
```suggestion
            public void messageReceived(ChannelHandlerContext ctx, ByteBuf msg) throws Exception {
                msg.release();
                SslHandler sslHandler = ctx.pipeline().get(SslHandler.class);
                SSLEngine engine = sslHandler.engine();
                logger.debug("Client message received: {} ({}) {} {}",
                        engine.getSession(), engine.getHandshakeStatus(), isSessionReused(engine),
                        Arrays.toString(engine.getSession().getValueNames()));
                Object value = engine.getSession().getValue("key");
                clientSessionValues.put((String) value);
            }
```

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 21]
```diff
@@ -3808,6 +3845,257 @@ public void checkServerTrusted(java.security.cert.X509Certificate[] chain,
...
+                .trustManager(serverTm)
+                .sslProvider(sslServerProvider())
+                .sslContextProvider(serverSslContextProvider())
+                .protocols(param.protocols())
+                .ciphers(param.ciphers())
+                .clientAuth(ClientAuth.REQUIRE)
+                .build());
+    }
+
+    private class SessionValueSettingTrustManager extends EmptyExtendedX509TrustManager
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java:4056]:
```suggestion
    private final class SessionValueSettingTrustManager extends EmptyExtendedX509TrustManager
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketSslSessionReuseTest.java [hunk 22]
```diff
@@ -213,4 +318,58 @@ public void exceptionCaught(ChannelHandlerContext ctx,
             ctx.close();
         }
     }
+
+    @SuppressJava6Requirement(reason = "Test only")
+    private static class SessionSettingTrustManager extends X509ExtendedTrustManager
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketSslSessionReuseTest.java:323]:
```suggestion
    private static final class SessionSettingTrustManager extends X509ExtendedTrustManager
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java (hunk 18 above)
Comment (normanmaurer) [reply handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:125]: hmm this is kind of stupid but ok... Maybe just wrap an `UnsupportedOperationException` in it then...

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 23]
```diff
@@ -667,7 +667,6 @@ private void rethrowIfFailed() {
         if (cause == null) {
             return;
         }
-
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:670]: just revert this one...

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+            //noinspection SuspiciousArrayCast
+            return (X509Certificate[]) peerCertificates;
+        }
+        X509Certificate[] chain = new X509Certificate[peerCertificates.length];
+        for (int i = 0; i < peerCertificates.length; i++) {
+            Certificate cert = peerCertificates[i];
+            if (cert instanceof X509Certificate) {
+                chain[i] = (X509Certificate) cert;
+            } else {
+                throw new IllegalArgumentException("Only X509Certificates are supported, found: " +
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:110]: should we just "skip" null and add it directly as we don't do any "null validation" if `peerCertificates` is of type `X509Certificate[]` ?

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+        return chain;
+    }
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    private static final class X509ExtendedWrapTrustManager extends X509ExtendedTrustManager {
+        private final X509ExtendedTrustManager trustManager;
+        private final Set<SSLEngine> confirmedValidations;
+
+        X509ExtendedWrapTrustManager(X509ExtendedTrustManager trustManager, Set<SSLEngine> confirmedValidations) {
+            this.trustManager = ObjectUtil.checkNotNull(trustManager, "trustManager");
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:123]: I think we can skip the null check here as we also don't do a null check for `confirmedValidations`.

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Almost there...

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java (hunk 25 above)
Comment (chrisvest) [reply handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:123]: Right, there's no way it can be null if it passed the instanceof check.

File: handler/src/main/java/io/netty/handler/ssl/ResumptionController.java [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+        if (peerCertificates instanceof X509Certificate[]) {
+            //noinspection SuspiciousArrayCast
+            return (X509Certificate[]) peerCertificates;
+        }
+        X509Certificate[] chain = new X509Certificate[peerCertificates.length];
+        for (int i = 0; i < peerCertificates.length; i++) {
+            Certificate cert = peerCertificates[i];
+            if (cert instanceof X509Certificate) {
+                chain[i] = (X509Certificate) cert;
+            } else {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ResumptionController.java:109]: This will throw if `cert == null` which it does not for the case when there is a null in it and its of type ` X509Certificate[]`. I think we should be consistent and just ignore a null value and put it in the array.

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
Great work... Ship it!

**Comment by chrisvest** [comment]
The Java 11 Graal build failure is caused by #14375.

The Java 11 BoringSSL build fails with an OOME in the HTTP/2 module `at io.netty.handler.codec.http2.HpackDynamicTable.setCapacity(HpackDynamicTable.java:182)`, and I'm not immediately convinced it's related to these changes.

**Comment by chrisvest** [comment]
Will do the forward merge tomorrow
